### PR TITLE
Allow using both define and module.exports.

### PR DESCRIPTION
**Pull request by ganemone**
If a script is loaded and uses require.js, module.exports will not be set. This means another script which uses commonjs will not be able to require punycode. This fixes that issue, allowing for both define + module.exports definitions.

### CODE REVIEW

**Comment by mathiasbynens** [comment]
[`tests/index.html`](https://github.com/bestiejs/punycode.js/blob/master/tests/index.html) tests require.js by default, and it works for me.

**Comment by ganemone** [comment]
I'm not sure how that is related. I described the situation in which this doesn't work above. Also, you can easily see the scenario where it would not setup module.exports in the code. 

> If a script is loaded and uses require.js, module.exports will not be set. This means another script which uses commonjs will not be able to require punycode.

The tests don't fall under this scenario.

**Comment by ganemone** [comment]
``
if (
		typeof define == 'function' &&
		typeof define.amd == 'object' &&
		define.amd
	) {
		define('punycode', function() {
			return punycode;
		});
	} else if (freeExports && freeModule) {
		if (module.exports == freeExports) {
			// in Node.js, io.js, or RingoJS v0.8.0+
			freeModule.exports = punycode;
		} else {
			// in Narwhal or RingoJS v0.7.0-
			for (key in punycode) {
				punycode.hasOwnProperty(key) && (freeExports[key] = punycode[key]);
			}
		}
	} else {
		// in Rhino or a web browser
		root.punycode = punycode;
	}
``

The issue is the "else if". If the check for amd passes, module.exports will not be set.

**Comment by mathiasbynens** [comment]
cc @jdalton

File: punycode.js [hunk 1]
```diff
@@ -507,15 +507,20 @@
 	/** Expose `punycode` */
 	// Some AMD build optimizers, like r.js, check for specific condition patterns
 	// like the following:
-	if (
```
Comment (jdalton) [inline punycode.js]: We should keep the AMD condition as is for the AMD build optimizers (see the js comment above).

**Comment by jdalton** [comment]
@ganemone Can you clarify the scenario (baby step me through it with first, second, third bits).

**Comment by jdalton** [comment]
I'm not sure when this will be an issue. Is it for code that's bundled with browserify/webpack and then loaded as an AMD module?

**Comment by ganemone** [comment]
So the current issue is this:

```
// This script sets up amd 
<script src="code/using/require.js"></script>
// This script tries to require('punycode')
<script src="code/using/commonjs"></script>
```

The second script is browserified. Since the amd check passes, module.exports will not be set. I'm realizing now however that maybe the solution is to use a browserify transform such as https://www.npmjs.com/package/amdclean.

**Comment by jdalton** [comment]
>  I'm realizing now however that maybe the solution is to use a browserify transform such as https://www.npmjs.com/package/amdclean.

Ah nice!

**Comment by jdalton** [comment]
I donno if this would get around this but in other projectsI export the lib to the global if run in the browser. Regardless and in addition to other AMD/CommonJS exports that may be in play.

**Comment by ganemone** [comment]
Yea actually that is not what I thought it was at first glance, haha. But I basically need a transform that removes the amd check. There is probably one out there. Closing for now

**Comment by jdalton** [comment]
Cool. Your scenario is an interesting one. I dig all the mix-and-mashups of it all.

**Comment by ganemone** [comment]
So turns out it doesn't look like there is an existing transform that removes the amd check. I'm wondering, could we simply change the ``else if`` into an ``if`` when checking for module.exports? Or if that will screw up the amd check, could we add a separate check at the end that simply checks for module.exports?

**Comment by mathiasbynens** [comment]
Punycode.js is now Node.js v6+ only.

**Pull request closed without merging**